battleship/battleship-multi.py

```diff
@@ -0,0 +1,320 @@
+                return False
+        # input ships and add to set
+        for i in range(startCoords[1], endCoords[1]+1):
+            shipCoords[shipIndex].add((startCoords[0], i))
+            # Print grid if specified (from player)
+            if grid != None:
+                grid[startCoords[0]][i] = 's'
+        return True
+    else:
+        # second coordinate is the same
+        if startCoords[0] > endCoords[0]:
+            startCoords, endCoords = endCoords, startCoords
+        # check validity to ensure coordinate doesn't already contain another ship
+        for i in range(startCoords[0], endCoords[0]+1):
+            if hitShip((i, startCoords[1]), shipCoords, shipIndex) >= 0:
+                return False
+        # input ships and add to set
+        for i in range(startCoords[0], endCoords[0]+1):
+            shipCoords[shipIndex].add((i, startCoords[1]))
+            # Print grid if specified (from player)
+            if grid != None:
+                grid[i][startCoords[1]] = 's'
+        return True
+
+
+def convert(coords):
+    '''Convert coords in the correct form to the 2D array coordinates'''
+
+    # Check correct length
+    if len(coords) == 2:
+        # Convert to ASCII
+        first = ord(coords[0].upper())
+        second = ord(coords[1])
+        # Check validity of coordinates
+        firstValid = first >= 65 and first <= 74
+        secondValid = second >= 48 and second <= 57
+        if firstValid and secondValid:
+            # Return appropriate tuple
+            return(first-65, second-48)
+
+
+def hitShip(coords, shipCoords, shipIndex=5):
+    '''Check if coordinate intersects with existing ship, and return the index of the ship if it does. If not, return -1'''
+
+    # Loop through ships
+    for i in range(shipIndex):
+        # Check if coordinate is in the set for specified ship
+        if coords in shipCoords[i]:
+            return i
+    return -1
+
+
+# Declare const variables
+shipNames = ['Carrier', 'Battleship', 'Cruiser', 'Submarine', 'Destroyer']
+shipLength = [5, 4, 3, 3, 2]
+
+# Full game loop
+while True:
+    # player2 view of player1's grid, initialize with 10x10 grid of ' '
+    playerOneGrid = [[' ' for i in range(10)] for j in range(10)]
+    # player1 view of player2's grid, initialize with 10x10 grid of ' '
+    playerTwoGrid = [[' ' for i in range(10)] for j in range(10)]
+    # Keep track of ships alive for player1 and player2 to determine when to end the game
+    playerOneShipsAlive = 5
+    playerTwoShipsAlive = 5
+    gameLength = 0
+    # Track coordinates for ships alive and sunk for both players in order to update char from 'o' to 'x' when the entire ship is sunk. Each index represents 1 ship.
+    playerOneShipAliveCoords = [set() for i in range(5)]
+    playerOneShipSunkCoords = [set() for i in range(5)]
+    playerTwoShipAliveCoords = [set() for i in range(5)]
+    playerTwoShipSunkCoords = [set() for i in range(5)]
```

Comment:
I’m pretty sure we haven’t taught list comprehension, so either replace them or we can add list comprehension to lesson 3

Reply:
Hmmm I feel like list comprehension is pretty useful to learn but it's also a little complicated, maybe we can just touch a little bit on it during the lesson? We can ask the channel

Reply:
Okay I think we'll go with the latter option then!